File: lec02/04_big_o/O1_constant.c
// O1_constant.c
#include <stdbool.h>
#include <stdio.h>

// O(1) - 상수 시간복잡도
// 입력 크기와 관계없이 항상 일정한 시간이 걸림

int get_first_element(int arr[], int size) {
  // 배열의 첫 번째 요소 반환 - O(1)
  if (size > 0) {
    return arr[0];
  }
  return -1;  // 빈 배열인 경우
}

int add_numbers(int a, int b) {
  // 두 숫자 더하기 - O(1)
  return a + b;
}

bool check_if_even(int n) {
  // 짝수인지 확인 - O(1)
  return n % 2 == 0;
}

int main() {
  int numbers[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  int size = sizeof(numbers) / sizeof(numbers[0]);

  printf("O(1) 예시들:\n");
  printf("첫 번째 요소: %d\n", get_first_element(numbers, size));
  printf("5 + 3 = %d\n", add_numbers(5, 3));
  printf("7은 짝수인가? %s\n", check_if_even(7) ? "예" : "아니오");
  printf("8은 짝수인가? %s\n", check_if_even(8) ? "예" : "아니오");

  return 0;
}

File: lec02/04_big_o/O1_constant.py
# O1_constant.py
# O(1) - 상수 시간복잡도
# 입력 크기와 관계없이 항상 일정한 시간이 걸림

def get_first_element(arr):
  """배열의 첫 번째 요소 반환 - O(1)"""
  if len(arr) > 0:
    return arr[0]
  return None


def add_numbers(a, b):
  """두 숫자 더하기 - O(1)"""
  return a + b


def check_if_even(n):
  """짝수인지 확인 - O(1)"""
  return n % 2 == 0


# 사용 예시
if __name__ == "__main__":
  numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

  print("O(1) 예시들:")
  print(f"첫 번째 요소: {get_first_element(numbers)}")
  print(f"5 + 3 = {add_numbers(5, 3)}")
  print(f"7은 짝수인가? {check_if_even(7)}")
  print(f"8은 짝수인가? {check_if_even(8)}")
